src/LinkedCellParticleContainer.cpp

```diff
@@ -0,0 +1,230 @@
+#include "../include/LinkedCellParticleContainer.h"
+
+#include <cmath>
+
+#include "spdlog/fmt/bundled/format.h"
+
+LinkedCellParticleContainer::LinkedCellParticleContainer(const std::array<double, 3>& domain_dims, double cutoff_radius,
+                                                         const std::array<BoundaryType, 6>& boundary_types)
+    : domainDims(domain_dims), cutoffRadius(cutoff_radius), boundaryTypes(boundary_types) {
+  if (cutoffRadius <= 0.0) {
+    throw std::invalid_argument("Cutoff radius must be positive");
+  }
+  initCells();
+}
+
+void LinkedCellParticleContainer::addParticle(std::array<double, 3> x, std::array<double, 3> v, double m) {
+  ParticleContainer::addParticle(x, v, m);
+  Particle& p = (*this)[this->size() - 1];  // the newly added particle
+  const int cdx = getCellIndex(p.getX());
+  cells[cdx].push_back(&p);
+}
+
+void LinkedCellParticleContainer::addParticle(const Particle* p) {
+  addParticle(p->getX(), p->getV(), p->getM());
+}
+
+void LinkedCellParticleContainer::updateCells() {
+  for (auto& cell : cells) {
+    cell.clear();
+  }
+
+  const std::size_t n = this->size();
+  for (std::size_t i = 0; i < n; ++i) {
+    Particle& p = (*this)[i];
+    const std::size_t c = getCellIndex(p.getX());
+    cells[c].push_back(&p);
+  }
+}
+
+LinkedCellParticleContainer::CellType LinkedCellParticleContainer::getCellType(size_t cdx) const {
+  const int nx = numCells[0];
+  const int ny = numCells[1];
+  const int nz = numCells[2];
+
+  const int idx = static_cast<int>(cdx);
+  const int layerSize = nx * ny;
+
+  const int iz = idx / layerSize;
+  const int remainder = idx % layerSize;
+  const int iy = remainder / nx;
+  const int ix = remainder % nx;
+
+  // outermost layer in any dimension -> Halo
+  if (ix == 0 || ix == nx - 1 || iy == 0 || iy == ny - 1 || iz == 0 || iz == nz - 1) {
+    return CellType::HALO;
+  }
+
+  // next layer  from halo -> Boundary
+  if (ix == 1 || ix == nx - 2 || iy == 1 || iy == ny - 2 || iz == 1 || iz == nz - 2) {
+    return CellType::BOUNDARY;
+  }
+
+  return CellType::INNER;
+}
+
+void LinkedCellParticleContainer::applyBoundaryConditions() {
+  // reflective boundaries are handled with ghost particles in ForceCalc
+  handleOutflow();
+  updateCells();
+}
+void LinkedCellParticleContainer::iteratePairs(const std::function<void(Particle&, Particle&)>& pairFunc) const {
+  const int nx = numCells[0];
+  const int ny = numCells[1];
+  const int nz = numCells[2];
+
+  const int layerSize = nx * ny;
+  // offsets for 13 forward neighbors
+  static const int neighborOffsets[13][3] = {{1, 0, 0},  {1, 1, 0},  {0, 1, 0},  {-1, 1, 0}, {-1, -1, 1},
```

Comment:
```suggestion
  static constexpr std::array<std::array<int, 3>13> neighborOffsets = {{1, 0, 0},  {1, 1, 0},  {0, 1, 0},  {-1, 1, 0}, {-1, -1, 1},
```
`constexpr` evaluates this array already at compile time. `std::array` is usually preferred, but in this case c style might be okay